### PR TITLE
Max Variable Size Checking

### PR DESCRIPTION
**Pull request by ndp-opendap**
This PR adds a new function `dap_utils::compute_response_size_and_inv_big_vars()` that inventories each constrained variable in the provided libdap object whose total size in bytes is in excess of a provided `max_size`. The call returns the total size, in bytes, of the constrained object(s).

The function is located in `bes/dap/DapUtils.cc` 

This is just the function for the calculation, not the application of it to determine if a request will be handled.  That is much more complex and I would like to separate that PR from this one.

### CODE REVIEW

**Review by jgallagher59701 — APPROVED** [review]
It would be good to add unit tests for get_dap_decl() and get_dap_array_dims_str(). 

But I think the general question was, is this correct. I think so.

File: dap/DapUtils.cc [hunk 1]
```diff
@@ -164,8 +177,137 @@ void throw_if_dap4_response_too_big(DMR &dmr, const std::string &file, unsigned
...
+/**
+ *
+ * @param grp
+ * @param max_var_size
+ * @param too_big
+ */
+uint64_t compute_response_size_and_inv_big_vars( libdap::D4Group *grp, const uint64_t &max_var_size, std::unordered_map<std::string,int64_t> &too_big)
+{
+    uint64_t response_size = 0;
+    auto cnstrctr = dynamic_cast<libdap::Constructor *>(grp);
```
[the remark this answers is not in the document]
Comment (ndp-opendap) [reply dap/DapUtils.cc:376]: > You can use a static_cast<>() to use 'grp' as a Constructor.

Is that preferable to `dynamic_cast<>()`?

File: dap/DapUtils.cc [hunk 2]
```diff
@@ -164,8 +177,137 @@ void throw_if_dap4_response_too_big(DMR &dmr, const std::string &file, unsigned
...
+                if(d4dim->c_stride()){
+                    double num = (elements+0.0)/d4dim->c_stride();
+                    elements  = num;
+                }
+            } else{
+                elements = d4dim->size();
+            }
+        }
+        else {
+            double num = dim.stop - (dim.start+0.0)/dim.stride;
```
[the remark this answers is not in the document]
Comment (ndp-opendap) [reply dap/DapUtils.cc]: done

File: dap/DapUtils.cc [hunk 3]
```diff
@@ -164,8 +177,137 @@ void throw_if_dap4_response_too_big(DMR &dmr, const std::string &file, unsigned
...
+            }
+            else {
+                BESDEBUG(MODULE, prolog << "ERROR Failed to cast BaseType pointer var to Constructor type." << endl);
+            }
+        }
+        else {
+            if (var->send_p()) {
+                uint64_t vsize = var->width_ll(true);
+                response_size += vsize;
+                string vdecl = get_dap_decl(var, true);
```
[the remark this answers is not in the document]
Comment (ndp-opendap) [reply dap/DapUtils.cc]: done

File: dap/unit-tests/DapUtilsTest.cc [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+
+
+/*##################################################################################################*/
+/* TESTS BEGIN */
+
+    void var_too_big_test() {
+
+        D4BaseTypeFactory d_d4f;
+        DMR *d_test_dmr;
+        d_test_dmr = new DMR(&d_d4f);
```
[the remark this answers is not in the document]
Comment (ndp-opendap) [reply dap/unit-tests/DapUtilsTest.cc:105]: I think the stack would ave been better, certainly simpler. But I did it with `unique_pointer` just to see.

[the remark this answers is not in the document]
Comment (ndp-opendap) [reply dap/unit-tests/DapUtilsTest.cc:105]: done

File: dap/unit-tests/DapUtilsTest.cc [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+        else {
+            CPPUNIT_FAIL("ERROR: No variables were deemed too big!");
+        }
+    }
+
+    void dmrpp_var_too_big_test() {
+
+        DMR *d_test_dmr;
+        D4BaseTypeFactory d_d4f;
+        d_test_dmr = new DMR(&d_d4f);
```
[the remark this answers is not in the document]
Comment (ndp-opendap) [reply dap/unit-tests/DapUtilsTest.cc:143]: ibid.

File: dap/DapUtils.cc [hunk 6]
```diff
@@ -164,8 +177,137 @@ void throw_if_dap4_response_too_big(DMR &dmr, const std::string &file, unsigned
...
+            elements  = num;
+            if(!elements) elements = 1;
+        }
+        my_dims << "[" << elements << "]";
+        dim_itr++;
+    }
+    return my_dims.str();
+}
+
+std::string get_dap_decl(libdap::BaseType *var, bool constrained=false) {
```
[the remark this answers is not in the document]
Comment (ndp-opendap) [reply dap/DapUtils.cc]: done

File: dap/DapUtils.cc [hunk 7]
```diff
@@ -164,8 +177,137 @@ void throw_if_dap4_response_too_big(DMR &dmr, const std::string &file, unsigned
         msg <<  " kilobyte response, which is too large. ";
         msg << "The maximum response size for this server is limited to " << dmr.response_limit_kb();
         msg << " kilobytes.";
-            throw BESSyntaxUserError(msg.str(), file, line);
+        throw BESSyntaxUserError(msg.str(), file, line);
+    }
+}
+
+std::string get_dap_array_dims_str(libdap::Array &a, bool constrained=false){
```
[the remark this answers is not in the document]
Comment (ndp-opendap) [reply dap/DapUtils.cc]: removed.

**Pull request merged**